### PR TITLE
Fix: event driven chain cache

### PR DESCRIPTION
**Pull request by LaurenceLiZhixin**
**What this PR does**:
This PR is based on pr #915
pr 915 aims at fix health check——add black list logic, and split conn check router from health check router.
This PR aims at using event-driven to let router send signal to notify channel, to notify chain to refresh its cache. 
Not ticker dirven like before, but event driven.
This can descrease the CPU usage, as there is no longer useless comparasion of invoker list of chain, but only compare when invokers/router change.

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: protocol/rpc_status.go [hunk 1]
```diff
@@ -181,4 +194,81 @@ func CleanAllStatus() {
 		return true
 	}
 	serviceStatistic.Range(delete2)
+	delete3 := func(key, _ interface{}) bool {
```
Comment (AlexStocks) [inline protocol/rpc_status.go:197]: delete3? can not get the meaning of this func name.

File: common/extension/conn_checker.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+)
+
+// SetHealthChecker sets the HealthChecker with @name
+func SetConnChecker(name string, fcn func(_ *common.URL) router.ConnChecker) {
+	connCheckers[name] = fcn
+}
+
+// GetHealthChecker gets the HealthChecker with @name
+func GetConnChecker(name string, url *common.URL) router.ConnChecker {
+	if connCheckers[name] == nil {
```
Comment (AlexStocks) [inline common/extension/conn_checker.go:36]: f, ok := connCheckers[mame]
         if !ok || f == nil {
		panic("connCheckers for " + name + " is not existing, make sure you have import the package.")
	}
	return connCheckers[name](url)

File: cluster/cluster_impl/base_cluster_invoker.go [hunk 3]
```diff
@@ -124,35 +124,43 @@ func (invoker *baseClusterInvoker) doSelectInvoker(lb cluster.LoadBalance, invoc
...
+		protocol.SetInvokerUnhealthyStatus(invokers[0])
+		logger.Errorf("the invokers of %s is nil. ", invocation.Invoker().GetUrl().ServiceKey())
+		return nil
 	}
 
 	selectedInvoker := lb.Select(invokers, invocation)
 
-	//judge to if the selectedInvoker is invoked
-
+	//judge to if the selectedInvoker is invoked and available
```
Comment (AlexStocks) [inline cluster/cluster_impl/base_cluster_invoker.go:139]: can not get the meaning of this chinglish sentence

File: protocol/rpc_status.go [hunk 4]
```diff
@@ -181,4 +194,81 @@ func CleanAllStatus() {
...
+}
+
+func GetAndRefreshState() bool {
+	state := blackListCacheDirty.Load()
+	blackListCacheDirty.Store(false)
+	return state
+}
+
+// TryRefreshBlackList start 3 gr to check at most block=16 invokers in black list
+// if is available remove from black list
```
Comment (AlexStocks) [inline protocol/rpc_status.go:248]: can  not know what is the meaning of this sentence.

File: cluster/router/conncheck/conn_check_route_test.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package conncheck
+
+import (
```
Comment (AlexStocks) [inline cluster/router/conncheck/conn_check_route_test.go:20]: pls split it.

**Review by AlexStocks — APPROVED** [review]

File: cluster/router/chain/chain_test.go [hunk 6]
```diff
@@ -148,7 +148,13 @@ conditions:
 	url := getConditionRouteUrl(applicationKey)
 	assert.NotNil(t, url)
 	factory := extension.GetRouterFactory(url.Protocol)
-	r, err := factory.NewPriorityRouter(url)
+	notify := make(chan struct{})
+	go func() {
```
[the remark this answers is not in the document]
Comment (LaurenceLiZhixin) [reply cluster/router/chain/chain_test.go:152]: fixed～

**Review by cityiron — APPROVED** [review]

File: cluster/cluster_impl/base_cluster_invoker.go [hunk 7]
```diff
@@ -121,38 +121,45 @@ func (invoker *baseClusterInvoker) doSelect(lb cluster.LoadBalance, invocation p
 
 func (invoker *baseClusterInvoker) doSelectInvoker(lb cluster.LoadBalance, invocation protocol.Invocation, invokers []protocol.Invoker, invoked []protocol.Invoker) protocol.Invoker {
 	if len(invokers) == 0 {
```
[the remark this answers is not in the document]
Comment (LaurenceLiZhixin) [reply cluster/cluster_impl/base_cluster_invoker.go:123]: 这个地方比较有意思，是改了个已有的坑。
invocation.Invoker()是个nil，所以一走到这一行就会panic。
后来发现这里不会被执行到。不存在invokers列表为空的情况。所以决定删掉这行日志了。

File: cluster/cluster_impl/base_cluster_invoker.go [hunk 8]
```diff
@@ -121,38 +121,45 @@ func (invoker *baseClusterInvoker) doSelect(lb cluster.LoadBalance, invocation p
...
 		return nil
 	}
+	go protocol.TryRefreshBlackList()
 	if len(invokers) == 1 {
-		return invokers[0]
+		if invokers[0].IsAvailable() {
+			return invokers[0]
+		}
+		protocol.SetInvokerUnhealthyStatus(invokers[0])
+		logger.Errorf("the invokers of %s is nil. ", invokers[0].GetUrl().ServiceKey())
```
[the remark this answers is not in the document]
Comment (LaurenceLiZhixin) [reply cluster/cluster_impl/base_cluster_invoker.go:132]: 这里位置主要是针对invokers列表因为健康检查导致为空时，打印出对应失败的key。

File: cluster/cluster_impl/base_cluster_invoker.go [hunk 9]
```diff
@@ -121,38 +121,45 @@ func (invoker *baseClusterInvoker) doSelect(lb cluster.LoadBalance, invocation p
...
+	//judge if the selected Invoker is invoked and available
 	if (!selectedInvoker.IsAvailable() && invoker.availablecheck) || isInvoked(selectedInvoker, invoked) {
+		protocol.SetInvokerUnhealthyStatus(selectedInvoker)
+		otherInvokers := getOtherInvokers(invokers, selectedInvoker)
 		// do reselect
-		var reslectInvokers []protocol.Invoker
-
-		for _, invoker := range invokers {
-			if !invoker.IsAvailable() {
+		for i := 0; i < 3; i++ {
```
[the remark this answers is not in the document]
Comment (LaurenceLiZhixin) [reply cluster/cluster_impl/base_cluster_invoker.go:143]: Yes, you r right, I've fixed it.

**Review by watermelo — APPROVED** [review]

**Pull request merged**